File: 03-RootFinding/2-regula-falsi.cpp
#include <iostream>
#include <cmath>

using fptr = double(double);
double f(double x);
double bisection(double xl, double xu, double eps, fptr fun);
double regulafalsi(double xl, double xu, double eps, fptr fun);

int main(int argc, char **argv)
{
    std::cout.precision(15); std::cout.setf(std::ios::scientific);
    double xl = 1;
    double xu = 20.0;
    double eps = 1.0e-12;

    double xroot = bisection(xl, xu, eps, f);
    std::cout << xroot << "\t" << f(xroot) << "\n";
    xroot = regulafalsi(xl, xu, eps, f);
    std::cout << xroot << "\t" << f(xroot) << "\n";

    return 0;
}

double f(double x)
{
    const double G = 9.81;
    const double M = 68.1;
    const double T = 10.0;
    const double VF = 40.0;
    return (M*G*(1 - std::exp(-x*T/M))/x) - VF;
}
double bisection(double xl, double xu, double eps, fptr fun)
{
    const int NITERMAX = 1000;
    double xr = 0;
    int niter = 0;
    while (niter <= NITERMAX) {
        xr = 0.5*(xl + xu);
        if (std::fabs(fun(xr)) <= eps) {
            break;
        } else if (fun(xr)*fun(xu) > 0) {
            xu = xr;
        } else {
            xl = xr;
        }
        niter++;
    }
    std::cout << "Bisection Info -> Niter: " << niter << "\n";
    return xr;
}

double regulafalsi(double xl, double xu, double eps, fptr fun)
{
    const int NITERMAX = 1000;
    double xr = 0;
    int niter = 0;
    while (niter <= NITERMAX) {
        double fxu = fun(xu);
        double fxl = fun(xl);
        xr = xu - (fxu*(xl-xu))/(fxl - fxu);
        if (std::fabs(fun(xr)) <= eps) {
            break;
        } else if (fun(xr)*fun(xu) > 0) {
            xu = xr;
        } else {
            xl = xr;
        }
        niter++;
    }
    std::cout << "RegulaFalsi Info -> Niter: " << niter << "\n";
    return xr;
}

File: 03-RootFinding/2-regula-falsi.py
import numpy as np

#Function whose root wants to be found
def f(x):
    g = 9.81
    m = 68.1
    T = 10.0
    VF = 40.0
    return m*g*(1-np.exp(-x*T/m))/x - VF

#Bisection algorithm
def bisection(xl, xu, eps, fun):
    NITERMAX = 1000
    xr = 0
    Niter = 0
    while Niter <= NITERMAX:
        xr = 0.5*(xl + xu)
        if abs(fun(xr)) < eps :
            break
        if fun(xr)*fun(xu) > 0:
            xu = xr
        else:
            xl = xr
        Niter+=1
    print("Bisection Info -> Niter: "  , Niter)
    return xr

#Regula-Falsi algorithm
def regula_falsi(xl, xu, eps, fun):
    NITERMAX = 1000
    xr = 0
    Niter = 0
    while Niter <= NITERMAX:
        fxu = fun(xu)
        fxl = fun(xl)
        xr = xu - (fxu*(xl - xu))/(fxl - fxu)
        if abs(fun(xr)) < eps :
            break
        if fun(xr)*fun(xu) > 0:
            xu = xr
        else:
            xl = xr
        Niter+=1
    print("Regula-Falsi Info -> Niter: "  , Niter)
    return xr

#Parameters to find the root
Xl = 1
Xu = 20
EPS = 1e-12

#Print roots and errors
xroot = bisection(Xl, Xu, EPS, f)
print(xroot, f(xroot))
xroot = regula_falsi(Xl, Xu, EPS, f)
print(xroot, f(xroot))



"""
//////////////////////////////////////////////////////////////////////////
Created on Fri Jan  1 13:32:11 2021                                    //
                                                                      //
@author: Nicolás Gómez                                               //
//////////////////////////////////////////////////////////////////////
"""

